src/parameterizations/vertical/MOM_set_viscosity.F90

```diff
@@ -1202,6 +1086,707 @@ subroutine set_viscous_BBL(u, v, h, tv, visc, G, GV, US, CS, pbv)
+  real :: L_2_reg          ! The value of L when vol_below is Vol_2_reg [nondim]
+  real :: vol_inflect_1    ! The volume at which there is an inflection point in the expression
+                           ! relating L to vol_err when there is a single open region [Z ~> m]
+  real :: vol_inflect_2    ! The volume at which there is an inflection point in the expression
+                           ! relating L to vol_err when there are two open regions [Z ~> m]
+
+  real :: L_inflect_1      ! The value of L that sits at an inflection point in the expression
+                           ! relating L to vol_err when there is a single open region [nondim]
+  real :: L_inflect_2      ! The value of L that sits at an inflection point in the expression
+                           ! relating L to vol_err when there is are two open regions [nondim]
+  real :: L_max, L_min     ! Maximum and minimum bounds on the solution for L for an interface [nondim]
+  real :: vol_err          ! The difference between the volume below an interface for a given value
+                           ! of L and the target value [Z ~> m]
+  real :: dVol_dL          ! The partial derivative of the volume below with L [Z ~> m]
+  real :: vol_err_max      ! The value of vol_err when L is L_max [Z ~> m]
+
+  ! The following combinations of slope and crv are reused across layers, and hence are pre-calculated
+  ! for efficiency.  All are non-negative.
+  real :: Icrvpslope       ! The inverse of the sum of crv and slope [Z-1 ~> m-1]
+  real :: slope_crv        ! The slope divided by the curvature [nondim]
+  ! These are only used if the slope exceeds or matches the curvature.
+  real :: smc              ! The slope minus the curvature [Z ~> m]
+  real :: C3c_m_s          ! 3 times the curvature minus the slope [Z ~> m]
+  real :: I_3c_m_s         ! The inverse of 3 times the curvature minus the slope [Z-1 ~> m-1]
+  ! These are only used if the curvature exceeds the slope.
+  real :: C4_crv           ! The inverse of a quarter of the curvature [Z-1 ~> m-1]
+  real :: sxcms_c          ! The slope times the difference between the curvature and slope
+                           ! divided by the curvature [Z ~> m]
+  real :: slope2_4crv      ! A quarter of the slope squared divided by the curvature [Z ~> m]
+  real :: I_3s_m_c         ! The inverse of 3 times the slope minus the curvature [Z-1 ~> m-1]
+  real :: C3s_m_c          ! 3 times the slope minus the curvature [Z ~> m]
+
+  real, parameter :: C1_3 = 1.0 / 3.0, C1_12 = 1.0 / 12.0 ! Rational constants [nondim]
+  integer :: K, nz, itt
+  integer, parameter :: max_itt = 10
+
+  nz = GV%ke
+
+  ! Each cell extends from x=-1/2 to 1/2, and has a topography
+  ! given by D(x) = crv*x^2 + slope*x + D_vel - crv/12.
+
+  crv_3 = (Dp + Dm - 2.0*D_vel) ; crv = 3.0*crv_3
+  if (crv <= 0.0) call MOM_error(FATAL, "find_L_open_concave should only be called with a positive curvature.")
+  slope = Dp - Dm
```

Comment:
Most comments about the `find_L_open_concave_trigonometric` apply equally here, except that `crv_3` could probably replace `crv` as the input argument, with `crv = 3.0 * crv_3` computed internally.

Again, the `if (crv <= 0)` test feels redundant.